pkg/services/ngalert/provisioning/alert_rules.go

```diff
@@ -0,0 +1,151 @@
+		interval = service.defaultInterval
+	} else if err != nil {
+		return models.AlertRule{}, err
+	}
+	rule.IntervalSeconds = interval
+	rule.Updated = time.Now()
+	err = service.xact.InTransaction(ctx, func(ctx context.Context) error {
+		ids, err := service.ruleStore.InsertAlertRules(ctx, []models.AlertRule{
+			rule,
+		})
+		if err != nil {
+			return err
+		}
+		if id, ok := ids[rule.UID]; ok {
+			rule.ID = id
+		} else {
+			return errors.New("couldn't find newly created id")
+		}
+		err = service.ruleStore.UpdateRuleGroup(ctx, rule.OrgID, rule.NamespaceUID, rule.RuleGroup, rule.IntervalSeconds)
+		if err != nil {
+			return err
+		}
+		return service.provenanceStore.SetProvenance(ctx, &rule, rule.OrgID, provenance)
+	})
+	if err != nil {
+		return models.AlertRule{}, err
+	}
+	return rule, nil
+}
+
+func (service *AlertRuleService) UpdateAlertRule(ctx context.Context, rule models.AlertRule, provenance models.Provenance) (models.AlertRule, error) {
+	storedRule, storedProvenance, err := service.GetAlertRule(ctx, rule.OrgID, rule.UID)
+	if err != nil {
+		return models.AlertRule{}, err
+	}
+	if storedProvenance != provenance && storedProvenance != models.ProvenanceNone {
+		return models.AlertRule{}, fmt.Errorf("cannot changed provenance from '%s' to '%s'", storedProvenance, provenance)
+	}
+	rule.Updated = time.Now()
+	rule.ID = storedRule.ID
+	rule.IntervalSeconds, err = service.ruleStore.GetRuleGroupInterval(ctx, rule.OrgID, rule.NamespaceUID, rule.RuleGroup)
+	if err != nil {
+		return models.AlertRule{}, err
+	}
+	service.log.Info("update rule", "ID", storedRule.ID, "labels", fmt.Sprintf("%+v", rule.Labels))
+	err = service.xact.InTransaction(ctx, func(ctx context.Context) error {
+		err := service.ruleStore.UpdateAlertRules(ctx, []store.UpdateRule{
+			{
+				Existing: &storedRule,
+				New:      rule,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		err = service.ruleStore.UpdateRuleGroup(ctx, rule.OrgID, rule.NamespaceUID, rule.RuleGroup, rule.IntervalSeconds)
+		if err != nil {
+			return err
+		}
+		return service.provenanceStore.SetProvenance(ctx, &rule, rule.OrgID, provenance)
+	})
+	if err != nil {
+		return models.AlertRule{}, err
+	}
+	return rule, err
+}
+
+func (service *AlertRuleService) DeleteAlertRule(ctx context.Context, orgID int64, ruleUID string, provenance models.Provenance) error {
+	rule := &models.AlertRule{
+		OrgID: orgID,
+		UID:   ruleUID,
+	}
+	// check that provenance is not changed in a invalid way
+	storedProvenance, err := service.provenanceStore.GetProvenance(ctx, rule, rule.OrgID)
+	if err != nil {
+		return err
+	}
+	if storedProvenance != provenance && storedProvenance != models.ProvenanceNone {
+		return fmt.Errorf("cannot delete with provided provenance '%s', needs '%s'", provenance, storedProvenance)
+	}
+	return service.xact.InTransaction(ctx, func(ctx context.Context) error {
+		err := service.ruleStore.DeleteAlertRulesByUID(ctx, orgID, ruleUID)
```

Comment:
If we delete the last rule in a group, should that also delete the group? (Maybe a thought for a future PR)

Reply:
Talked about this offline, groups are part of the rules for now in the DB schema. So when the last rule of the group is deleted, it removes automatically the group as it's part of the rule record.